src/turing.jl

```diff
@@ -0,0 +1,83 @@
+function from_turing(
+    chns=nothing;
+    model=nothing,
+    rng=Random.default_rng(),
+    nchains=ndraws = chns isa Turing.MCMCChains.Chains ? last(size(chns)) : 1,
+    ndraws=chns isa Turing.MCMCChains.Chains ? first(size(chns)) : 1_000,
+    library=Turing,
+    observed_data=nothing,
+    constant_data=nothing,
+    posterior_predictive=nothing,
+    prior=nothing,
+    prior_predictive=nothing,
+    log_likelihood=nothing,
+    kwargs...,
+)
+    groups = Dict{Symbol,Any}(
+        :observed_data => observed_data,
+        :constant_data => constant_data,
+        :posterior_predictive => posterior_predictive,
+        :prior => prior,
+        :prior_predictive => prior_predictive,
+        :log_likelihood => log_likelihood,
+    )
+    model === nothing && return from_mcmcchains(chns; library=library, groups..., kwargs...)
+    if groups[:prior] === nothing
+        groups[:prior] = reduce(
+            Turing.chainscat,
+            map(
+                _ -> Turing.sample(rng, model, Turing.Prior(), ndraws; progress=false),
+                1:nchains,
+            ),
+        )
+    end
+
+    groups[:observed_data] === nothing &&
+        return from_mcmcchains(chns; library=library, groups..., kwargs...)
+
+    observed_data = groups[:observed_data]
+    data_var_names = Set(
+        observed_data isa Dict ? Symbol.(keys(observed_data)) : propertynames(observed_data)
+    )
+
+    if groups[:constant_data] === nothing
+        groups[:constant_data] = NamedTuple(
+            filter(p -> first(p) ∉ data_var_names, pairs(model.args))
+        )
+    end
+
+    # Instantiate the predictive model
+    args_pred = NamedTuple(
+        k => k in data_var_names ? similar(v, Missing) : v for (k, v) in pairs(model.args)
+    )
+    model_predict = Turing.DynamicPPL.Model(model.name, model.f, args_pred, model.defaults)
```

Comment:
Unfortunately this won't work for a couple of different reasons.

`similar(v, Missing)` is not always going to be the right thing to do due to limitations with how Turing "discovers" whether something should be observed or not. E.g. in the case of `x ~ MvNormal(10, 1.0)`, you'd need to make `x = missing` rather than `x = fill(missing, 10)`, while if `x[i] ~ Normal()` for every `i`, then you'd need to make `x = fill(missing, 10)` rather than `x = missing`. The following snippet demonstrates what I mean:
```julia
using DynamicPPL, Distributions

@model function demo(x, n = x isa AbstractArray ? length(x) : 1)
    s ~ InverseGamma(2, 3)
    m ~ Normal(0, √s)
    for i in 1:n
        x[i] ~ Normal(m, √s)
    end

    return x
end

# (✓) Considers `x` to be `missing` and therefore samples a new value for it.
demo(fill(missing, 2))()
# 2-element Vector{Union{Missing, Float64}}:
#  -1.1754445514231586
#  -3.5132773181923156

# (×) Doesn't see `x` as `missing` and will error trying to compute `logpdf` with `missing`.
demo(missing)()
# MethodError: no method matching setindex!(::Missing, ::Float64, ::Int64)

@model function demo_vec(x, n = x isa AbstractArray ? length(x) : 1)
    s ~ InverseGamma(2, 3)
    m ~ Normal(0, √s)
    x ~ MvNormal(m * ones(n), √s)
end
# (✓) Considers `x` to be `missing` and therefore samples a new value for it.
demo_vec(missing, 2)()
# 2-element Vector{Float64}:
#  -2.8388279305946194
#  -1.2962882427789901

# (×) Doesn't see `x` as `missing` and will error trying to compute `logpdf` with `missing`.
demo_vec(fill(missing, 2))()
# MethodError: no method matching loglikelihood(::IsoNormal, ::Vector{Union{Missing, Float64}})
```
But it's possible to obtain more control over this by manually instantiating a `DynamicPPL.Model` instance as follows:

```julia
#######################
### Manual override ###
#######################
# A `DynamicPPL.Model` has the following definition:
# 
#   Model{F,argnames,defaultnames,missings,Targs,Tdefaults}
#
# and so we can actually override which variables are considered `missing` using the constructor
#
#  Model{missings}(name, f, args::NamedTuple, defaults::NamedTuple)
#
# NOTE: This has the drawback that we can't specify that only a _part_ of an vector is `missing`, e.g.
# `x[1] isa Real` and `x[2] === missing` cannot be specified this way; only the full vector `x` can be
# set to `missing`.

# NOTE: So what's considered missing in a model is the union of:
# 1. The `missings` in `Model`.
# 2. What we can infer as `missing` at runtime, e.g. `fill(missing, 2)` ONLY when the `missing` part
#    corresponds to how it's access in the model, e.g. `x ~ MvNormal()` requires `x === missing` while
#    `x[i] ~ Normal()` requires `x[i] === missing`.

m = demo(fill(missing, 2))
# Notice how the `missing` of the `Model` is just `()` in the below, despite the fact that `x` will be
# treated as `missing` when we run the model.
typeof(m)
# Model{var"#39#40", (:x, :n), (:n,), (), Tuple{Vector{Missing}, Int64}, Tuple{Int64}}
DynamicPPL.getmissings(m)
# ()
m()
# 2-element Vector{Union{Missing, Float64}}:
#  -0.8793252702256413
#   0.5390169363224118

# This one doesn't work.
m = demo_vec(fill(missing, 2))
m()
# MethodError: no method matching loglikelihood(::IsoNormal, ::Vector{Union{Missing, Float64}})

# So we manually tell `Model` that `x` is actually missing and should be sampled:
m2 = Model{(:x, )}(m.name, m.f, m.args, m.defaults)
m2()
# 2-element Vector{Float64}:
#  -1.7832640955001318
#  -0.4884815968096884

# When you specify it in the `Model` type, you don't even need to `fill` with missing!
m3 = Model{(:x, )}(m.name, m.f, merge(m.args, (x = zeros(2), )), m.defaults)
m3()
# 2-element Vector{Float64}:
#  -1.071192960480266
#  -0.9565956686160692
```

Soooo this finally brings us to the solution which should work for all cases:

```suggestion
model_predict = Turing.DynamicPPL.Model{data_var_names}(model.name, model.f, args_pred, model.defaults)
```